File: apps/backend/services/hybridRetrieve.ts
import { qdrantClient } from "@repo/qdrant/client";
import { prismaClient } from "@repo/prisma/client";
import { getDenseVectors, getSparseVectors } from "@repo/embed/client";
import { startActiveObservation, truncateForTrace } from "@repo/langfuse/client";

const COLLECTION = process.env.COLLECTION as string;
const DEFAULT_RETRIEVAL_LIMIT = 50;

export type RetrievedChunk = {
    id: string;
    text: string;
    score: number;
    documentId?: string | number | null;
    documentTitle?: string | null;
    tags?: string[] | null;
    modality?: string | null;
};

export type HybridRetrieveOptions = {
    /** Max RRF points to return (default 50). Use higher values for document search aggregation. */
    limit?: number;
    modality?: string;
};

/**
 * Hybrid retrieval: dense (cosine) + sparse (SPLADE), fused with RRF in Qdrant.
 * Scoped to the requesting user's chunks only.
 */
export async function hybridRetrieve(
    userId: string,
    query: string,
    options: HybridRetrieveOptions | string = {}
): Promise<RetrievedChunk[]> {
    // Back-compat: chat used to pass modality as the third positional arg
    const opts: HybridRetrieveOptions = typeof options === "string" ? { modality: options } : options;
    const limit = Math.min(Math.max(opts.limit ?? DEFAULT_RETRIEVAL_LIMIT, 1), 200);
    const modality = opts.modality;

    return startActiveObservation(
        "hybrid-retrieve",
        async (retriever) => {
            retriever.update({
                input: { userId, query: truncateForTrace(query, 500) },
                metadata: {
                    collection: COLLECTION,
                    limit,
                    fusion: "rrf",
                    modality: modality ?? null,
                },
            });

            const ownedDocuments = await prismaClient.document.findMany({
                where: { userId },
                select: { id: true },
            });
            const ownedDocumentIds = ownedDocuments.map((doc) => doc.id);
            if (ownedDocumentIds.length === 0) {
                console.log(`[hybridRetrieve] No documents for userId=${userId}, skipping retrieval`);
                retriever.update({ output: { chunkCount: 0, reason: "no-documents" } });
                return [];
            }

            const mustConditions: Record<string, unknown>[] = [ { key: "documentId", match: { any: ownedDocumentIds } } ];
            if (modality) mustConditions.push({ key: "modality", match: { value: modality } });
            
            const filter = { must: mustConditions };

            console.log(`[hybridRetrieve] Starting retrieval for userId=${userId}, query: "${query.slice(0, 120)}"`);
            const [denseVectors, sparseVectors] = await Promise.all([
                getDenseVectors([query]),
                getSparseVectors([query]),
            ]);
            console.log(
                `[hybridRetrieve] Dense vector dims: ${denseVectors[0]?.length ?? 0}, Sparse vector nnz: ${sparseVectors[0]?.indices?.length ?? 0}`
            );

            const denseVector = denseVectors[0];
            const sparse = sparseVectors[0];

            if (!denseVector || !sparse) {
                console.error("[hybridRetrieve] Failed to embed query — no vectors returned");
                throw new Error("Failed to embed query");
            }

            const rawIndices = Array.from(sparse.indices as Iterable<number>);
            const rawValues = Array.from(sparse.values as Iterable<number>);
            const paired = rawIndices
                .map((idx, i) => ({ idx, val: rawValues[i] ?? 0 }))
                .filter((p) => p.val !== 0)
                .sort((a, b) => a.idx - b.idx);
            const sparseQuery = {
                indices: paired.map((p) => p.idx),
                values: paired.map((p) => p.val),
            };

            const denseQuery = Array.from(denseVector as ArrayLike<number>);

            if (denseQuery.some((v) => !Number.isFinite(v))) {
                console.error("[hybridRetrieve] Dense vector contains NaN or Infinity");
                throw new Error("Dense vector contains invalid values");
            }
            if (sparseQuery.indices.length === 0) {
                console.error("[hybridRetrieve] Sparse vector has no non-zero entries");
                throw new Error("Sparse vector is empty");
            }

            console.log(
                `[hybridRetrieve] Querying Qdrant collection "${COLLECTION}" with dense (${denseQuery.length}d) + sparse (${sparseQuery.indices.length} nnz) RRF, limit=${limit}, userId=${userId}, documents=${ownedDocumentIds.length}`
            );
            let res;
            try {
                res = await qdrantClient.query(COLLECTION, {
                    prefetch: [
                        {
                            query: denseQuery,
                            using: "dense",
                            limit,
                            filter,
                        },
                        {
                            query: sparseQuery,
                            using: "splade",
                            limit,
                            filter,
                        },
                    ],
                    query: { fusion: "rrf" },
                    limit,
                    filter,
                    with_payload: true,
                });
            } catch (qdrantErr: any) {
                console.error(`[hybridRetrieve] Qdrant query failed:`, {
                    message: qdrantErr.message,
                    status: qdrantErr.status,
                    statusText: qdrantErr.statusText,
                    data: JSON.stringify(qdrantErr.data),
                });
                throw qdrantErr;
            }

            const rawChunks = (res.points ?? [])
                .map((point) => {
                    const payload = (point.payload ?? {}) as Record<string, unknown>;
                    const tagsRaw = payload.tags;
                    const tags = Array.isArray(tagsRaw)
                        ? tagsRaw.filter((t): t is string => typeof t === "string")
                        : null;
                    return {
                        id: String(point.id),
                        text: typeof payload.text === "string" ? payload.text : "",
                        score: point.score ?? 0,
                        documentId: (payload.documentId as string | number | undefined) ?? null,
                        documentTitle:
                            typeof payload.documentTitle === "string"
                                ? payload.documentTitle
                                : null,
                        tags,
                        modality:
                            typeof payload.modality === "string" ? payload.modality : null,
                        payloadUserId:
                            typeof payload.userId === "string" ? payload.userId : null,
                    };
                })
                .filter((c) => c.text.length > 0);

            const ownedDocumentIdSet = new Set(ownedDocumentIds);

            const chunks = rawChunks
                .filter((c) => {
                    const docId = c.documentId == null ? null : String(c.documentId);
                    if (!docId || !ownedDocumentIdSet.has(docId)) return false;
                    return c.payloadUserId == null || c.payloadUserId === userId;
                })
                .map(({ payloadUserId: _payloadUserId, ...chunk }) => chunk);

            console.log(
                `[hybridRetrieve] Qdrant returned ${res.points?.length ?? 0} RRF points, ${chunks.length} user-scoped chunks with text`
            );

            retriever.update({
                output: {
                    chunkCount: chunks.length,
                    topScores: chunks.slice(0, 5).map((c) => ({
                        id: c.id,
                        score: c.score,
                    })),
                },
            });

            return chunks;
        },
        { asType: "retriever" }
    );
}

File: apps/api/app/services/hybrid_retrieve.py
from __future__ import annotations

import logging
import re
from typing import Any

from qdrant_client import QdrantClient
from qdrant_client.http import models as qm

from app.config import get_settings
from app.db import get_pool
from app.services.embed import get_dense_vectors, get_sparse_vectors

logger = logging.getLogger("recallos.retrieve")

TAG_MATCH_BOOST_PER_TAG = 0.04
TAG_MATCH_BOOST_CAP = 0.12


def _tag_match_boost(query: str, tags: list[str] | None) -> float:
    if not tags:
        return 0.0
    q = query.lower().strip()
    if not q:
        return 0.0
    query_tokens = set(t for t in re.split(r"[^a-z0-9]+", q) if len(t) >= 2)
    boost = 0.0
    for raw in tags:
        tag = raw.strip().lower()
        if not tag:
            continue
        if tag in q:
            boost += TAG_MATCH_BOOST_PER_TAG
            continue
        tag_tokens = [t for t in re.split(r"[^a-z0-9]+", tag) if len(t) >= 2]
        if tag_tokens and all(tt in query_tokens for tt in tag_tokens):
            boost += TAG_MATCH_BOOST_PER_TAG * 0.75
            continue
        if len(tag_tokens) == 1 and tag_tokens[0] in query_tokens:
            boost += TAG_MATCH_BOOST_PER_TAG
    return min(boost, TAG_MATCH_BOOST_CAP)


def _qdrant() -> QdrantClient:
    s = get_settings()
    return QdrantClient(host=s.qdrant_host, port=s.qdrant_port)


async def hybrid_retrieve(
    user_id: str,
    query: str,
    *,
    limit: int = 50,
    modality: str | None = None,
) -> list[dict[str, Any]]:
    settings = get_settings()
    limit = max(1, min(limit, 200))

    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            'SELECT id FROM "Document" WHERE "userId" = $1', user_id
        )
    owned = [str(r["id"]) for r in rows]
    if not owned:
        return []

    dense = get_dense_vectors([query])[0]
    sparse = get_sparse_vectors([query])[0]
    if not sparse["indices"]:
        raise RuntimeError("Sparse vector is empty")

    must: list[qm.FieldCondition] = [
        qm.FieldCondition(key="documentId", match=qm.MatchAny(any=owned))
    ]
    if modality:
        must.append(
            qm.FieldCondition(key="modality", match=qm.MatchValue(value=modality))
        )
    filt = qm.Filter(must=must)

    client = _qdrant()
    res = client.query_points(
        collection_name=settings.collection,
        prefetch=[
            qm.Prefetch(
                query=dense,
                using="dense",
                limit=limit,
                filter=filt,
            ),
            qm.Prefetch(
                query=qm.SparseVector(
                    indices=sparse["indices"], values=sparse["values"]
                ),
                using="splade",
                limit=limit,
                filter=filt,
            ),
        ],
        query=qm.FusionQuery(fusion=qm.Fusion.RRF),
        limit=limit,
        query_filter=filt,
        with_payload=True,
    )

    owned_set = set(owned)
    chunks: list[dict[str, Any]] = []
    for point in res.points or []:
        payload = point.payload or {}
        text = payload.get("text") or ""
        if not text:
            continue
        doc_id = payload.get("documentId")
        doc_id_s = str(doc_id) if doc_id is not None else None
        if not doc_id_s or doc_id_s not in owned_set:
            continue
        payload_uid = payload.get("userId")
        if payload_uid is not None and str(payload_uid) != user_id:
            continue
        tags = payload.get("tags")
        tags_list = [t for t in tags if isinstance(t, str)] if isinstance(tags, list) else []
        score = float(point.score or 0) + _tag_match_boost(query, tags_list)
        chunks.append(
            {
                "id": str(point.id),
                "text": text,
                "score": score,
                "documentId": doc_id_s,
                "documentTitle": payload.get("documentTitle"),
                "tags": tags_list,
                "modality": payload.get("modality"),
                "page": payload.get("page"),
                "timestampStart": payload.get("timestampStart"),
                "timestampEnd": payload.get("timestampEnd"),
                "caption": payload.get("caption"),
                "chunkId": payload.get("chunkId"),
            }
        )
    chunks.sort(key=lambda c: c["score"], reverse=True)
    return chunks
